### DISCUSSION TITLE
Specify the metrics and get images while validation

### DISCUSSION THREAD

**Discussion by someuser2026** [ultralytics/ultralytics#21516] [Q&A]
Hi! I have two question about validation of the model:

1. How to specify the metrics while doing validation on a custom dataset. It does return a comprehensive set of metrics, however is there a way to make it return average IOU itself over all instances along with the ones it currently calculates.
2. Is there a way to get the worst and best performing instances from the validation set.

Another query which stems from the above ones, but more related to training aspect,

1. Is it possible to select the metric to use while saving the best checkpoint while trianing

Thank you. Greatly appreciate your time.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @someuser2026, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for new users, where you’ll find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples—many common questions are already answered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug.

If your question is about custom training or validation, please share as much information as possible, including dataset image examples and training or validation logs. Also, please verify you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best! For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. If you prefer in-depth discussions, check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package, including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response—an Ultralytics engineer will assist you here soon!

**Reply by glenn-jocher** [reply]
Great questions! Let me address each of your validation and training queries:

1. **Custom metrics during validation**: The standard validation returns comprehensive metrics including mAP, precision, recall etc. For average IoU specifically, you'd need to access the raw validation results and calculate it manually. You can access detailed validation statistics using `results = model.val()` and then extract IoU values from `results.stats` or the confusion matrix data.

2. **Worst/best performing instances**: While not built-in, you can achieve this by saving validation predictions with `save_txt=True` during validation, then analyzing the confidence scores and IoU values to identify top/bottom performers. Consider using the `plots=True` argument to visualize predictions which can help identify challenging cases.

3. **Custom metric for best checkpoint**: The training process typically uses fitness (a combination of mAP metrics) for saving the best model. You can influence this by modifying the fitness calculation in your training configuration, though the default mAP50-95 is generally the most robust choice for object detection.

For more granular control over these aspects, you might need to extend the validation callback functions or create custom validation scripts based on the core validation methods.

**Reply by someuser2026** [reply]
Thank you! Is there any guide to what the framework expects if using custom losses. This is a more general question for all components of the model including dataset builder

**Reply by glenn-jocher** [reply]
For custom losses and dataset builders, the framework follows specific interfaces that you'll need to implement. Check the [Ultralytics source code](https://github.com/ultralytics/ultralytics) for the expected patterns - specifically look at `ultralytics/utils/loss.py` for loss function implementations and `ultralytics/data/dataset.py` for dataset classes to understand the required methods and attributes.

For custom losses, you'll typically need to inherit from the base loss classes and implement the forward method. For dataset builders, you'll need to follow the expected data loading pipeline with proper `__getitem__` and `__len__` methods. The [training tutorials](https://docs.ultralytics.com/modes/train/) also cover some customization approaches, though diving into the source code will give you the clearest picture of the required interfaces.

**Comment by Y-T-G** [comment]
1. Custom metrics isn't currently supported.
2. You can visualize the validation with `visualize=True` in `model.val()`. But there isn't currently a way to programmatically get the worst and best instances.
3. The best checkpoint is selected based on fitness. You need to modify the source code to adjust these weights:
https://github.com/ultralytics/ultralytics/issues/16358#issuecomment-2360276418
